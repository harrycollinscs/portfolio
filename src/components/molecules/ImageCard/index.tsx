import React, { ReactElement } from 'react'
import styled, { withTheme } from 'styled-components'

interface Props {
  image: { src: string, alt: string }
  headerLeft: ReactElement
  headerRight: ReactElement
  title?: ReactElement
  paragraphs: string[]
  isList?: boolean
  link?: string
  theme?: any 
}

const Card = styled.div`
position: relative;
  align-content: center;
  background-color: ${({ theme }) => theme.container};
  line-height: 2;
  text-align: center;
  width: 100%;
  border-radius: 5px;
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;

  img {
    border-radius: 5px 5px 0px 0px;
    width: 100%;
    height: 200px;
    object-fit: cover;
    overflow: hidden; 

    @media screen and (max-width: 700px) {
      border-radius: 0;
    }
  }

  div {
    padding: 24px;
  }

  button {
    border: 0;
    margin: 20px 20px;
    padding: 10px 20px;
    background-color:  ${({ theme }) => theme.accentBackground};
    color: ${({ theme }) => theme.accentText};
    border-radius: 5px;
    position: absolute;
    bottom: 0;
    right: 0;
  }

  @media screen and (max-width: 700px) {
    border-radius: 0;
  }
`

const List = (paras: string[]) => (
  <ul style={{ textAlign: 'left' }}>
    {paras.map(para => <li>{para}</li>)}
  </ul>
)

const Paragraphs = (paras: string[]) => (
  <>
  {paras.map((para: string) => (
        <p>{para}</p>
      ))}
  </>
)

const ImageCard = ({ image, headerLeft, headerRight, title, paragraphs, isList, link }: Props) => (
  <Card>
    <img src={image.src} alt={image.alt}/>
    <div>
      <div style={{ display: 'flex', justifyContent: 'space-between', textAlign: 'center', alignItems: 'center' }}>
        {headerLeft}
        {headerRight}
      </div>
      {!!title && title}
      { isList ? List(paragraphs) : Paragraphs(paragraphs) }
    </div>
    {
      !!link && (
        <div style={{ width: '100%', height: '40px' }}>
          <a href={link} target="_blank" rel="noreferrer">
            <button >See more</button>
          </a>
        </div>
      )
    }
  </Card>
)

export default withTheme(ImageCard)