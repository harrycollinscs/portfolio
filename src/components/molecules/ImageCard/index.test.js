import React from 'react'
import renderer from 'react-test-renderer'
import ImageCard from './index'
import PortfolioLight from '../../../assets/images/PortfolioLightMode.png'

const testParagraphs = [
  'I recently began this project as a means of self-development. With the goal of increasing my breadth of knowledge in the world of Software Development, I have created a Notion document containing some important topics that I believe will be helpful to learn for any Full Stack Engineer.',
  'Each topic has a variety of sub-topics/key areas to go into more detail with.',
  'The development of this resource has been both for my own learning, and for me to congregate the knowledge I already have. Thus, some content has been written by me directly, and some has been with the assistance of ChatGPT, and then reviewed before addition.'
]

it('renders correctly', () => {
  const tree = renderer.create(
    <ImageCard
      image={{
        src: PortfolioLight,
        alt: 'portfolio'
      }}
      headerLeft={<p>Test left</p>}
      headerRight={<p>Test right</p>}
      paragraphs={testParagraphs}
    />
  ).toJSON()
  expect(tree).toMatchSnapshot()
});
