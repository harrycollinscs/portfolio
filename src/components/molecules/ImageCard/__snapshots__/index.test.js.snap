// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-bcXHqe kIdong"
>
  <img
    alt="portfolio"
    src={Object {}}
  />
  <div>
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "space-between",
          "textAlign": "center",
        }
      }
    >
      <p>
        Test left
      </p>
      <p>
        Test right
      </p>
    </div>
    <p>
      I recently began this project as a means of self-development. With the goal of increasing my breadth of knowledge in the world of Software Development, I have created a Notion document containing some important topics that I believe will be helpful to learn for any Full Stack Engineer.
    </p>
    <p>
      Each topic has a variety of sub-topics/key areas to go into more detail with.
    </p>
    <p>
      The development of this resource has been both for my own learning, and for me to congregate the knowledge I already have. Thus, some content has been written by me directly, and some has been with the assistance of ChatGPT, and then reviewed before addition.
    </p>
  </div>
</div>
`;
